### PR TITLE
fix(opencode): align agency-swarm stream lifecycle with session processor

### LINKED ISSUE #2: Fix Agency Swarm Responses stream translation regressions
## Problem
Agency Swarm stream translation introduced regressions in the OpenCode fork:
- duplicated terminal text output when both `response.output_text.*` and `message_output_created` are emitted for the same message
- teardown could be skipped on some adapter/nested error paths due to early returns, leaving turns without consistent terminal step events
- reasoning summaries could be reopened/duplicated during `response.output_item.done`

## Scope
- maintain OpenCode session/event contracts while using Agency transport endpoints
- preserve proper finish-step/finish semantics
- add regression tests for text dedupe, teardown-on-error, and reasoning finalization

## Acceptance
- no duplicate terminal output
- all error paths preserve teardown and final events
- no reasoning part reopen/duplicate on output-item finalization

### PR DESCRIPTION
**Pull request by nicko-ai**
Fixes #2

## Summary
This PR aligns Agency Swarm turns with OpenCode’s standard session processing path and fixes lifecycle bugs that caused duplicated or re-opened output.

## What Changed
- routes `agency-swarm` execution through `SessionProcessor.process()` via `createStream`, instead of provider-specific direct persistence
- translates Agency stream frames (including Responses-style events) into standard processor stream parts (`text-*`, `reasoning-*`, `tool-*`, `finish-step`, `finish`)
- removes now-unused direct-write helper module (`session/agency-swarm-parts.ts`) and centralizes state transitions in the stream translator
- preserves teardown ordering on all error paths before surfacing terminal error events

## Bug Fixes Included
- fixes duplicate terminal text when both `response.output_text.*` and `run_item_stream_event.message_output_created` are emitted for the same item
- prevents reasoning part reopen/duplication after already-closed summaries
- fixes finish-reason loop bug: completed Agency turns now end as `stop` (not `tool-calls`)
- fixes AbortError classification: external aborts now finalize as `cancelled` without false `error`
- marks dangling unfinished tool calls as terminal `error` to avoid silent incomplete state

## Scope and Intent
- keeps Agency-specific transport concerns in the adapter layer (Agency endpoints, run-id cancellation, required `chat_history` payload shape)
- keeps OpenCode session/event/storage contracts unchanged for UI compatibility
- does not add new backend endpoints

## Tests
Updated `packages/opencode/test/session/agency-swarm.test.ts` with regression coverage for:
- text/reasoning/tool lifecycle mapping
- duplicate text prevention
- teardown on adapter/nested errors
- no reasoning reopen on `response.output_item.done`
- completed-turn finish reason (`stop`)
- external abort behavior (`cancelled`, no spurious error)
- dangling tool call termination (`error`)

Validation run:
- `bun run typecheck` (packages/opencode)
- `bun test test/session/agency-swarm.test.ts`
- `bun test test/agency-swarm/adapter.test.ts`
- `bun test test/session/llm.test.ts`
- `bun test test/session`

## Upstream Sync Scope
This branch intentionally limits shared OpenCode surface-area changes to integration hooks and lifecycle correctness:
- `session/prompt.ts`: Agency branch now runs through `SessionProcessor` via injected stream
- `session/processor.ts`: minimal `createStream` hook to reuse standard processor lifecycle
- `session/index.ts`: Agency cost/cache metadata normalization
- `session/agency-swarm.ts`: Agency-specific stream translator/state machine
- tests/docs: Agency regressions and fork integration invariants

No backend API expansion was introduced.

### CODE REVIEW

File: packages/opencode/src/session/agency-swarm.ts [hunk 1]
```diff
@@ -213,269 +747,582 @@ export namespace SessionAgencySwarm {
...
+            }
+            continue
+          }
+
+          if (frame.type === "error") {
+            yield {
+              type: "error",
+              error: new Error(frame.error),
+            }
+            return
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply packages/opencode/src/session/agency-swarm.ts]: Fixed in 8a1ff4213.

I removed early returns on adapter error paths so teardown always runs: `flushOpen()`, `finish-step`, and `finish` are emitted before surfacing the error event. I also added a regression test (`stream preserves teardown before surfacing adapter error`) to lock this behavior.

**Comment by nicko-ai** [comment]
@codex review

File: packages/opencode/src/session/agency-swarm.ts [hunk 2]
```diff
@@ -213,269 +747,583 @@ export namespace SessionAgencySwarm {
...
+                })
+                continue
+              }
+
+              if (itemType === "reasoning") {
+                const itemID = asString(item["id"]) || lastReasoningItemID
+                if (!itemID) continue
+                for (const key of Array.from(reasoningByItem.get(itemID) ?? [])) {
+                  const index = Number(key.split(":")[1] || "0")
+                  yield* finishReasoning(itemID, Number.isFinite(index) ? index : 0, undefined, eventMeta, {
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply packages/opencode/src/session/agency-swarm.ts]: Fixed in 41b9a5b7e.

I prevented reasoning reopens by (1) not reopening closed keys in `finishReasoning` when there is no new content, (2) removing keys from per-item tracking when they are closed, and (3) only finalizing currently-open keys on `response.output_item.done`. Added regression test coverage for this sequence.

File: packages/opencode/src/session/agency-swarm.ts [hunk 3]
```diff
@@ -213,269 +747,583 @@ export namespace SessionAgencySwarm {
...
               continue
             }
+
+            if (responseType === "error") {
+              const message = asString(nested["message"]) || asString(nested["error"]) || "Unknown stream error"
+              yield {
+                type: "error",
+                error: new Error(message),
+              }
+              return
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply packages/opencode/src/session/agency-swarm.ts]: Fixed in 41b9a5b7e.

I changed `raw_response_event` `type: "error"` handling to set `streamError` and break, so teardown still runs (`flushOpen`, `finish-step`, `finish`) before emitting the final error event.

**Comment by nicko-ai** [comment]
@codex review

**Comment by nicko-ai** [comment]
Addressed the two open Codex findings in 861137221:\n\n- Fixed AbortError classification: plain AbortError now marks the stream as aborted/cancelled (not error), so turns don’t get incorrectly failed.\n- Fixed finish reason semantics: completed Agency turns no longer return `tool-calls`; they return `stop` unless cancelled/error.\n- Added guard for similar edge case: unfinished/dangling tool calls now finalize as `error` (with tool-error + terminal error event) to prevent follow-up loop duplication.\n\nAdded/updated tests in `test/session/agency-swarm.test.ts` for all three behaviors.

**Comment by nicko-ai** [comment]
@codex review

File: packages/opencode/src/session/prompt.ts [hunk 4]
```diff
@@ -644,29 +651,33 @@ export namespace SessionPrompt {
...
+              abort,
+              registerManagedCancel(handler) {
+                setManagedCancel(sessionID, handler)
+              },
+              clearManagedCancel() {
+                setManagedCancel(sessionID, undefined)
+              },
+            }),
         })
         break
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply packages/opencode/src/session/prompt.ts]: Fixed in f43cc89f9.

The Agency branch now captures `processor.process(...)` result and handles it like the standard path:
- `stop` -> break
- `compact` -> `SessionCompaction.create(...)`
- otherwise continue

Added regression test `test/session/prompt-agency-swarm.test.ts` that forces compact return and asserts compaction scheduling.

**Comment by nicko-ai** [comment]
@codex review

File: packages/opencode/src/session/agency-swarm.ts [hunk 5]
```diff
@@ -213,269 +1008,327 @@ export namespace SessionAgencySwarm {
...
-            if (responseType === "response.output_text.delta") {
+            if (responseType === "response.output_text.delta" || responseType === "response.refusal.delta") {
               const delta = asRawString(nested["delta"])
               if (delta === undefined) continue
+              const itemID = textItemID(nested)
+              if (!itemID) continue
+              const contentIndex = asNumber(nested["content_index"]) ?? textIndex.get(itemID) ?? 0
+              textIndex.set(itemID, contentIndex)
+              const textMeta = outputMeta(outputIndex, { content_index: contentIndex })
+              yield* ensureText(itemID, contentIndex, eventMeta, textMeta)
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply packages/opencode/src/session/agency-swarm.ts]: Fixed in 619780550.

Root cause was `textIndex` being updated before `ensureText(...)`, which prevented closing the active part on `content_index` transitions.

Changes:
- Removed pre-`ensureText` `textIndex.set(...)` writes in stream handlers.
- `ensureText(...)` now computes/closes the active key from prior state first.
- `finishText(...)` now avoids reopening already-closed parts when `final` is absent/equal (prevents duplicate empty text segments on `output_item.done` after `output_text.done`).

Added regression test `stream closes prior text part before switching content_index` in `test/session/agency-swarm.test.ts`.

**Comment by nicko-ai** [comment]
@codex review

File: packages/opencode/src/session/agency-swarm.ts [hunk 6]
```diff
@@ -213,269 +1018,325 @@ export namespace SessionAgencySwarm {
...
-          end: Date.now(),
+      yield {
+        type: "finish",
+      }
+
+      if (streamError) {
+        yield {
+          type: "error",
+          error: streamError,
         }
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply packages/opencode/src/session/agency-swarm.ts]: Fixed in 1bb9d4520.

Root cause confirmed: emitting `finish-step`/`finish` before `error` allowed `SessionProcessor` to persist step-finish/cost before retry handling.

Change: when `streamError` is set, the stream now emits terminal `error` immediately after teardown (`flushOpen`) and returns without `finish-step`/`finish`. This keeps retryable failures from creating duplicate finished steps/cost accounting.

Updated tests:
- `stream preserves teardown before surfacing adapter error`
- `stream preserves teardown when raw_response_event emits type=error`
- `stream marks unfinished tool calls as error instead of tool-calls`

File: packages/opencode/src/session/agency-swarm.ts [hunk 7]
```diff
@@ -213,269 +1018,325 @@ export namespace SessionAgencySwarm {
...
+                  output_index: outputIndex,
+                  item_type: itemType,
+                  phase,
+                })
+                continue
+              }
+
+              if (phase === "completed") {
+                yield* completeTool(callID, toolName, toolOutput(itemType, item), eventMeta, {
+                  item_id: itemID,
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply packages/opencode/src/session/agency-swarm.ts:1260]: Fixed in 1bb9d4520.

Root cause confirmed: generic `*_call.completed` path prematurely called `completeTool(...)` for `function_call`, which could suppress later real `function_call_output`.

Change: `response.function_call.completed` now only transitions to running (`runTool`) and does not complete the tool; completion is deferred to output events/messages.

Added regression test: `stream does not complete function_call on response.function_call.completed before output`.

**Comment by nicko-ai** [comment]
Addressed two additional root-cause bugs in `14a28ebb4`.

1. Prevented premature non-function tool completion on `response.*_call.completed`.
- `*_call.completed` now advances lifecycle to running only.
- Final completion stays sourced from `response.output_item.done`, `run_item_stream_event.tool_output`, or `messages` reconciliation.
- This avoids dropping real outputs that arrive later on `response.output_item.done` (verified against agency-staging fixtures for file/web/mcp calls).

2. Fixed stale-cancel behavior from sidecar `last_run_id`.
- Cancel no longer targets previous-turn `last_run_id` before current run metadata arrives.
- Added cancel dispatch when `run_id` first appears in `messages` payload (meta-missing path).

Regression coverage updated in `packages/opencode/test/session/agency-swarm.test.ts`:
- non-function completion waits for `response.output_item.done` and preserves real payload
- cancel-before-meta does not cancel stale run IDs

Validation run:
- `bun test test/session/agency-swarm.test.ts`
- `bun test test/session/prompt-agency-swarm.test.ts`
- `bun run typecheck`
- `bun test test/session`

**Comment by nicko-ai** [comment]
@codex review

File: packages/opencode/src/session/agency-swarm.ts [hunk 8]
```diff
@@ -213,269 +1021,328 @@ export namespace SessionAgencySwarm {
...
 
-      if (textPart && !textPart.time?.end) {
-        textPart.time = {
-          start: textPart.time?.start ?? Date.now(),
-          end: Date.now(),
+      if (streamError) {
+        yield {
+          type: "error",
+          error: streamError,
         }
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply packages/opencode/src/session/agency-swarm.ts:1315]: Fixed in 74fadebf1.

I kept the stream-level behavior (`error` emitted without `finish-step`) to avoid duplicate step-finish accounting on retryable failures, and fixed the root cause in `SessionProcessor` instead: non-retry terminal errors now set `assistantMessage.finish = "error"` in catch before persisting the message.

This preserves finish semantics for loop/UI state without re-introducing duplicate finish-step/cost persistence.

Added regression coverage in `test/session/processor-custom-stream.test.ts`:
- `marks message finish as error when custom stream emits terminal error without finish-step`

File: packages/opencode/src/session/processor.ts [hunk 9]
```diff
@@ -42,15 +42,21 @@ export namespace SessionProcessor {
...
         const shouldBreak = (await Config.get()).experimental?.continue_loop_on_deny !== true
         while (true) {
           try {
             let currentText: MessageV2.TextPart | undefined
             let reasoningMap: Record<string, MessageV2.ReasoningPart> = {}
-            const stream = await LLM.stream(streamInput)
+            const stream = streamInput.createStream ? await streamInput.createStream() : await LLM.stream(streamInput)
 
             for await (const value of stream.fullStream) {
               input.abort.throwIfAborted()
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply packages/opencode/src/session/processor.ts]: Fixed in 74fadebf1.

Root cause confirmed. `SessionProcessor` was checking `abort.throwIfAborted()` before consuming each frame, which could skip terminal cancellation frames from custom streams.

Change:
- moved abort enforcement to post-frame handling and only for non-custom streams (`!createStream`), so custom streams can deliver terminal `finish-step(cancelled)`/`finish` frames.

Added regression coverage in `test/session/processor-custom-stream.test.ts`:
- `processes cancelled finish-step from custom stream when abort is already set`

**Comment by nicko-ai** [comment]
@codex review

**Comment by nicko-ai** [comment]
@codex review

File: packages/opencode/src/session/agency-swarm.ts [hunk 10]
```diff
@@ -156,21 +156,829 @@ export namespace SessionAgencySwarm {
...
+        const parts = runTool(
+          callID,
+          toolName,
+          eventMeta,
+          outputMeta(outputIndex, {
+            item_type: itemType,
+          }),
+        )
+        if (itemType === "function_call") {
+          return parts
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply packages/opencode/src/session/agency-swarm.ts:873]: Fixed in afcfa0814.\n\nRoot cause confirmed: `response.output_item.done` for `*_call` items was not reconciling final input payload before `runTool(...)`, so `tool.raw` could stay stale when `output_item.added` had partial/empty args and no `response.*_arguments.done` arrived.\n\nChange:\n- Reconcile `toolRawInput(itemType, item)` from `output_item.done` before `runTool(...)` for all `*_call` items.\n- Keep `function_call` completion semantics unchanged (still completes on output/messages, not on `function_call` done).\n\nRegression tests added in `test/session/agency-swarm.test.ts`:\n- `stream uses final function_call arguments from response.output_item.done before tool-call`\n- `stream reconciles non-function tool input from response.output_item.done`\n\nValidation:\n- `bun test test/session/agency-swarm.test.ts`\n- `bun run typecheck`\n- `bun test`

**Comment by nicko-ai** [comment]
@codex review

**Comment by nicko-ai** [comment]
Fixed stale recipient-agent routing in `e16803322` with live metadata validation per turn.

What changed:
- `SessionAgencySwarm.stream` now validates recipient candidates (message mention first, then configured fallback) against fresh `get_metadata` every turn before sending `recipient_agent`.
- Stale configured recipients are skipped instead of causing backend `422 invalid agent name` failures.
- Agent labels from metadata are resolved to canonical ids (alias -> id).
- Prompt `@` autocomplete now refetches Agency discovery while the menu is open, so renamed/new agents appear without restarting the CLI.

Regression coverage added:
- stale configured recipient is skipped
- configured alias resolves to live id
- stale mentioned recipient falls back to valid configured recipient

Validation:
- `bun test test/session/agency-swarm.test.ts`
- `bun run typecheck`
- `bun test` (full `packages/opencode`)

**Comment by nicko-ai** [comment]
@codex review
